File: 8. String to Integer (atoi)/string_to_integer_atoi.js
function myAtoi(s) {
  const INT_MIN = -2147483648;
  const INT_MAX = 2147483647;

  s = s.trimStart();
  if (s.length === 0) return 0;

  let i = 0;
  let sign = 1;

  if (s[i] === '-' || s[i] === '+') {
    sign = (s[i] === '-') ? -1 : 1;
    i++;
  }

  let num = 0;
  while (i < s.length && s[i] >= '0' && s[i] <= '9') {
    num = num * 10 + (s[i].charCodeAt(0) - '0'.charCodeAt(0));
    i++;
  }

  num *= sign;

  if (num < INT_MIN) return INT_MIN;
  if (num > INT_MAX) return INT_MAX;
  return num;
}

function main() {
  console.log(myAtoi("42"));             // 42
  console.log(myAtoi("   -042"));        // -42
  console.log(myAtoi("1337c0d3"));       // 1337
  console.log(myAtoi("0-1"));            // 0
  console.log(myAtoi("words and 987")); // 0
}

main();

File: 008. String to Integer (atoi)/string_to_integer_atoi.py
def myAtoi(s):
    INT_MIN = -2**31
    INT_MAX = 2**31 - 1

    s = s.lstrip()
    if not s:
        return 0

    sign = 1
    i = 0

    if s[i] == '-' or s[i] == '+':
        sign = -1 if s[i] == '-' else 1
        i += 1

    num = 0

    while i < len(s) and s[i].isdigit():
        num = num * 10 + int(s[i])
        i += 1

    num *= sign

    if num < INT_MIN:
        return INT_MIN
    if num > INT_MAX:
        return INT_MAX

    return num

if __name__ == "__main__":
    print(myAtoi("42"))           # 42
    print(myAtoi("   -042"))      # -42
    print(myAtoi("1337c0d3"))     # 1337
    print(myAtoi("0-1"))          # 0
    print(myAtoi("words and 987")) # 0
